ClientRuntimes/Python/msrestazure/msrestazure/azure_active_directory.py

```diff
@@ -216,7 +228,70 @@ def clear_cached_token(self):
             raise_with_traceback(KeyError, "Unable to clear token.")
 
 
-class UserPassCredentials(OAuthTokenAuthentication, AADMixin):
+class AADRefreshMixin(object):
+    """
+    Additional token refresh logic
+    """
+
+    def refresh_session(self):
+        """Return updated session if token has expired, attempts to
+        refresh using newly acquired token.
+
+        :rtype: requests.Session.
+        """
+        if self.token.get('refresh_token'):
+            try:
+                return self.signed_session()
+            except Expired:
+                pass
+        self.set_token()
+        return self.signed_session()
+
+
+class AADTokenCredentials(AADMixin):
+    """
+    Credentials objects for AAD token retrieved through external process
+    e.g. Python ADAL lib.
+
+    Optional kwargs may include:
+    - china (bool): Configure auth for China-based service,
```

Comment:
Instead of taking the baseUri as a parameter in the constructor, it would be nice, if the user can provide the name of the environment. Default should be public azure. The generated code should have the base uri like this:
```
function StorageManagementClient(credentials, options) {
   . . .
   . . . .
  if (!options.environment) {
    options.environment = msrestAzure.AzureEnvironment.Azure;
  }
  this.environment = options.environment;
  this.baseuri = this.environment.resourceManagerEndpointUrl;
}